llvm/lib/Target/AMDGPU/AMDGPUISelDAGToDAG.cpp

```diff
@@ -2096,21 +2099,47 @@ bool AMDGPUDAGToDAGISel::SelectSMRDBaseOffset(SDValue Addr, SDValue &SBase,
   }
   if (!N0 || !N1)
     return false;
+
+  bool Selected = false;
   if (SelectSMRDOffset(N1, SOffset, Offset, Imm32Only, IsBuffer)) {
     SBase = N0;
-    return true;
-  }
-  if (SelectSMRDOffset(N0, SOffset, Offset, Imm32Only, IsBuffer)) {
+    Selected = true;
+  } else if (SelectSMRDOffset(N0, SOffset, Offset, Imm32Only, IsBuffer)) {
     SBase = N1;
+    Selected = true;
+  }
+
+  if (Selected) {
+    // For unbuffered smem loads, it is illegal and undefined for the Immediate
+    // Offset to be negative if the resulting (Offset + (M0 or SOffset or zero)
+    // is negative. Handle the case where the Immediate Offset is negative and
+    // there is no SOffset.
+    //
+    // FIXME: Also handle M0 or SOffset case?
```

Comment:
Not sure what the fixme is about; I also doubt the rules are different if it's m0 or a numbered SGPR

Reply:
We need to check if:
ImmOffset + (Sgpr[Offset] or M0) > 0
However, the value of sgpr[offset] or m0 may not be known at compile time so we cannot guarantee the above rule will not be broken. In my patch we are only currently checking if ImmOffset > 0. We still need to handle sgpr[offset] and m0 but I am not certain how to do that right now.